SpeechResponder/eddi.json

```diff
@@ -33,7 +33,7 @@
 reportbody.landable :\r\n        It has a\r\n        {if reportbody.rotationalperiod < -20.0:\r\n            slow retrograde\r\n        |elif reportbody.rotationalperiod < -0.5:\r\n            retrograde\r\n        |elif reportbody.rotationalperiod < 0:\r\n            fast retrograde\r\n        |elif reportbody.rotationalperiod <= 0.5:\r\n            fast\r\n        |elif reportbody.rotationalperiod > 20.0:\r\n            slow\r\n        }\r\n        rotational period of {Humanise(reportbody.rotationalperiod)}\r\n        day{if Humanise(reportbody.rotationalperiod) != \"1\":s}.\r\n\r\n        {set goodmaterials to []}\r\n        {set greatmaterials to []}\r\n        {for material in reportbody.materials:\r\n            {if material.percentage >= MaterialDetails(material.material).greatpctbody:\r\n                {set greatmaterials to cat(greatmaterials, [material])}\r\n            |elif material.percentage >= MaterialDetails(material.material).goodpctbody:\r\n                {set goodmaterials to cat(goodmaterials, [material])}\r\n            }\r\n        }\r\n\r\n        {if len(greatmaterials) > 0:\r\n            This body contains very high levels of\r\n            {if len(greatmaterials) = 1:\r\n                {greatmaterials[0].material}\r\n            |elif len(greatmaterials) = 2:\r\n                {greatmaterials[0].material} and {greatmaterials[1].material}\r\n            |else:\r\n                {set cur to 0}\r\n                {while cur < len(greatmaterials):\r\n                    {if cur = 0:\r\n                        {greatmaterials[cur].material}\r\n                    |elif cur < len(greatmaterials) - 1:\r\n                        , {greatmaterials[cur].material}\r\n                    |else:\r\n                        , and {greatmaterials[cur].material}\r\n                    }\r\n                    {set cur to cur + 1}\r\n                }\r\n            }\r\n        }\r\n        {if len(goodmaterials) > 0:\r\n            {if len(greatmaterials) > 0:\r\n                and high levels of\r\n            |else:\r\n                This body contains high levels of\r\n            }\r\n            {if len(goodmaterials) = 1:\r\n                {goodmaterials[0].material}\r\n            |elif len(goodmaterials) = 2:\r\n                {goodmaterials[0].material} and {goodmaterials[1].material}\r\n            |else:\r\n                {set cur to 0}\r\n                {while cur < len(goodmaterials):\r\n                    {if cur = 0:\r\n                        {goodmaterials[cur].material}\r\n                    |elif cur < len(goodmaterials) - 1:\r\n                        , {goodmaterials[cur].material}\r\n                    |else:\r\n                        , and {goodmaterials[cur].material}\r\n                    }\r\n                    {set cur to cur + 1}\r\n                }\r\n            }\r\n        }\r\n        {if len(greatmaterials) > 0 || len(goodmaterials) > 0:\r\n            .\r\n        }\r\n    }\r\n}",
```

Comment:
Looks good, but I think jgm intended this to say `{P(reportbody.name)} is a ...` so that if queried from context the body name could still be announced. (with your current revision, a VoiceAttack query from context would not give the body's name)
Rather than always using `This is a ...`, maybe we can add an additional if statement conditional on whether the script was triggered from a scan? (event.name will be null if the script wasn't triggered from a scan)
e.g. `{if event.name: This |else: {P(state.eddi_context_body_shortname)} }`

Reply:
Ah, I wasn't aware of the VA usage. Thanks. Will revise tomorrow.

Reply:
In case the Body report script gets called from a context where `state.eddi_context_body_shortname` is not set, I went with:

```
    {if event.name:
        This
    |else: 
        {P(default(state.eddi_context_body_shortname, reportbody.name))} 
    }
```

Reply:
No good. The variable `state.eddi_context_body_shortname` shall only be set if `event.name` is already set. Since the state variable is not cleared after use, if called from context this will use the shortname from the last scan event rather than the true name from current context. Once a shortname is set, `reportbody.name` will not be used again in that session.

Reply:
Suggested revision, incorporating many tweaks from @Darkcyde13's script and a few of my own enhancements:
```
{_ Set reportbody to event data _}
{set reportbody to event}
{set eventdata to true}

{_ Fallback to BodyDetails if no event data _}
{if !reportbody.name || reportbody.name = "":
    {set reportbody to BodyDetails(state.eddi_context_body_name, state.eddi_context_body_system)}
    {set eventdata to false}
}

{_ Second fallback if BodyDetails fails with system argument above _}
{if !reportbody.name || reportbody.name = "":
    {set reportbody to BodyDetails(state.eddi_context_body_name)}
    {set eventdata to false}
}

{_ Begin the report }
{if !reportbody.name || reportbody.name = "":
    {_ If it's still not possible to get the body name...  _}
    I'm
    {OneOf("not sure","unsure")}
    which body
    {OneOf("you are","you're")}
    {OneOf("asking about","referring to", "referencing")}.
|else:
    {if event.name:
        {if state.eddi_context_body_shortname: 
            {set shortname to Occasionally(3, state.eddi_context_body_shortname)} }
        {if !shortname: This |else: {shortname} }
    |else: 
        {reportbody.name} 
    }
    is
    {if reportbody.gravity:
        {if reportbody.gravity < 0.5:
            a low-gravity
        |elif reportbody.gravity < 2:
            a medium-gravity
        |elif reportbody.gravity < 4:
            a high-gravity
        |else:
            an extremely high-gravity
        }
    }

    {if reportbody.tidallylocked:
        tidally-locked
    }

    {if reportbody.terraformstate = "Terraformable":
        terraformable
    |elif reportbody.terraformstate = "Terraformed":
        terraformed
    }

    {if !reportbody.bodyclass: {set bodyclass to planettype} }
    {if find(reportbody.bodyclass, "Class") >= 0 && find(reportbody.bodyclass, "gas giant") >= 0:
        {set sudarsky to Occasionally(2, "Sud-dar-ski")}
        {if find(reportbody.bodyclass, "Class V") >= 0:
            {Occasionally(2, "{Sudarsky} class five")}
        |elif find(reportbody.bodyclass, "Class IV") >= 0:
            {Occasionally(2, "{Sudarsky} class four")}
        |elif find(reportbody.bodyclass, "Class III") >= 0:
            {Occasionally(2, "{Sudarsky} class three")}
        |elif find(reportbody.bodyclass, "Class II") >= 0:
            {Occasionally(2, "{Sudarsky} class two")}
        |elif find(reportbody.bodyclass, "Class I") >= 0:
            {Occasionally(2, "{Sudarsky} class one")}
        }
        gas giant
    |elif reportbody.bodyclass:
        {reportbody.bodyclass}
    |else: 
        body
    }

    {if len(reportbody.rings):
        with {if len(reportbody.rings) = 1: a}
        {if find(reportbody.reserves, "Pristine") >= 0:
            pristine
        |elif find(reportbody.reserves, "Major") >= 0:
            rich
        }
        {set ringname to []}
        {set ringcomposition to []}
        {for Ring in reportbody.rings:
            {set ringcomposition to cat(ringcomposition, [Ring.composition])}
        }
        {if len(ringcomposition) > 0:
            {List(ringcomposition)}
        }
        {if find(ringname[0], "Belt") >= 0: 
            {set ring to "belt"}
        |else:
            {set ring to "ring"}
        }
        {ring}{if len(reportbody.rings) > 1 || len(event.rings) > 1:s}
    }

    {if reportbody.terraformstate = "Terraforming":
        in the process of being terraformed
    }

    {if type(reportbody.landable) != "void": 
        that is
        {if reportbody.landable:
            suitable
        |else:
            unsuitable
        }
        for landing
    }
    . {_ note the full stop}

    {_ Atmosphere? }

    {if reportbody.volcanism:
        This planet shows signs of volcanism, with
        {if reportbody.volcanism.type = 'Geysers':
            {if reportbody.volcanism.amount = 'Major':
                high numbers of
            |elif reportbody.volcanism.amount = 'Minor':
                low numbers of
            }
            active {reportbody.volcanism.composition} geysers
        |else:
            {if reportbody.volcanism.amount = 'Major':
                high levels of
            |elif reportbody.volcanism.amount = 'Minor':
                low levels  of
            }
            active {reportbody.volcanism.composition} magma flows
        }.
    }

    {if reportbody.landable :
        It has a
        {if reportbody.rotationalperiod < -20.0:
            slow retrograde
        |elif reportbody.rotationalperiod < -0.5:
            retrograde
        |elif reportbody.rotationalperiod < 0:
            fast retrograde
        |elif reportbody.rotationalperiod <= 0.5:
            fast
        |elif reportbody.rotationalperiod > 20.0:
            slow
        }
        rotational period of {Humanise(reportbody.rotationalperiod)}
        day{if Humanise(reportbody.rotationalperiod) != "1":s}.

        {set goodmaterials to []}
        {set greatmaterials to []}
        {for material in reportbody.materials:
            {if material.percentage >= MaterialDetails(material.material).greatpctbody:
                {set greatmaterials to cat(greatmaterials, [material])}
            |elif material.percentage >= MaterialDetails(material.material).goodpctbody:
                {set goodmaterials to cat(goodmaterials, [material])}
            }
        }

        {if len(greatmaterials) > 0:
            This body contains very high levels of
            {if len(greatmaterials) = 1:
                {greatmaterials[0].material}
            |elif len(greatmaterials) = 2:
                {greatmaterials[0].material} and {greatmaterials[1].material}
            |else:
                {set cur to 0}
                {while cur < len(greatmaterials):
                    {if cur = 0:
                        {greatmaterials[cur].material}
                    |elif cur < len(greatmaterials) - 1:
                        , {greatmaterials[cur].material}
                    |else:
                        , and {greatmaterials[cur].material}
                    }
                    {set cur to cur + 1}
                }
            }
        }
        {if len(goodmaterials) > 0:
            {if len(greatmaterials) > 0:
                and high levels of
            |else:
                This body contains high levels of
            }
            {if len(goodmaterials) = 1:
                {goodmaterials[0].material}
            |elif len(goodmaterials) = 2:
                {goodmaterials[0].material} and {goodmaterials[1].material}
            |else:
                {set cur to 0}
                {while cur < len(goodmaterials):
                    {if cur = 0:
                        {goodmaterials[cur].material}
                    |elif cur < len(goodmaterials) - 1:
                        , {goodmaterials[cur].material}
                    |else:
                        , and {goodmaterials[cur].material}
                    }
                    {set cur to cur + 1}
                }
            }
        }
        {if len(greatmaterials) > 0 || len(goodmaterials) > 0:
            .
        }
    }
}
```
Note that the versions @Darkcyde13 and I use also describe the body's radius and mass, but that may be a little too chatty for most?